### PR TITLE
feat: support for inheritance default field values

### PR DESCRIPTION
**Pull request by tomwwinter**
related: #2217

### Visible/Frontend Changes
- [ ] When creating/editing an entity, configured inheritance default values are set

### Architectural/Backend Changes
- [ ] created separated service for default-field-values
- [ ] created use case for default-field-values logic

### CODE REVIEW

**Review by sleidig — CHANGES_REQUESTED** [review]
Really nice, clean code :heart_eyes:   Thanks for adding JsDoc also.

Functional testing:
- [x] static default value should be applied on new record
- [x] dynamic default value ($now, $current_user) should be applied on new record
- [ ] dynamic default value should not overwrite existing value :warning: 
- [x] static default value should not overwrite (explicitly cleared) empty field on existing entity :warning: 
- [x] inherited default value should be applied whenever the related reference is changed (on new/existing entity)
- [ ] inherited value can be overwritten manually :warning: (currently resets to inherited upon save)

maybe we can add some of these edge cases to the unit tests, also? (-> TDD)

File: src/app/core/entity/default-field-value.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+    // then
+    expect(
+      mockHandleDefaultFieldValuesUseCase.handleFormGroup,
+    ).not.toHaveBeenCalled();
+
+    Entity.schema.delete("test");
+  });
+
+  it("should apply inheritance modes before static and dynamic modes", () => {
```
[the remark this answers is not in the document]
Comment (tomwwinter) [reply src/app/core/entity/default-field-value.service.spec.ts:79]: The Inheritance mode listen to changes and loads the entities then. When static value is processed, before the inheritance hook is registered, the loading does not trigger.

File: src/app/core/entity/default-field-value/handle-default-field-values.usecase.spec.ts [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+import { LoggingService } from "../../logging/logging.service";
+import { FormBuilder, FormControl } from "@angular/forms";
+import { EntitySchemaField } from "../schema/entity-schema-field";
+import { Entity } from "../model/entity";
+
+function getDefaultInheritanceFormGroup() {
+  return new FormBuilder().group({
+    "field-1": new FormControl(),
+    "field-2": new FormControl(),
+    "reverence-1": new FormControl(),
```
[the remark this answers is not in the document]
Comment (tomwwinter) [reply src/app/core/entity/default-field-value/handle-default-field-values.usecase.spec.ts:15]: yes 🙈

File: src/app/core/entity/default-field-value/handle-default-field-values.usecase.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+      formGroup.get(fieldName);
+
+    if (!sourceFormControl || !targetFormControl) {
+      return;
+    }
+
+    sourceFormControl.valueChanges.subscribe(async (change) => {
+      if (targetFormControl.dirty && !!targetFormControl.value) {
+        return;
+      }
```
[the remark this answers is not in the document]
Comment (tomwwinter) [reply src/app/core/entity/default-field-value/handle-default-field-values.usecase.ts:73]: We need to separate the "isNew" and the "editing" case here, when creating new, we want to overwrite the value, if empty and not dirty.

when we edit a field, we don't want to overwrite a value, even when it is untouched. Right?

**Review by sleidig — COMMENTED** [review]
Functional testing:
| | static | dynamic | inherited |
| -------- | -------- | ------- | ------- |
| correct default value applied on new record                               | x | x | x (upon selecting ref) |
| can overwrite manually while creating new                                 | x | x | x |
| can delete/clear manually while creating new                             | x | x | x |
| keeps overwritten value when opening / editing existing             | x | x | x |
| keeps empty overwritten value when opening / editing existing  | x | x | x |

inherited value functionalities:
- [x] sets inherited values upon selecting the reference (new entity)
- [x] updates inherited values upon changing reference
- [x] clears inherited values upon removing reference
- [x] keeps manually set values, not overwriting them even when reference changes
- [ ] applies inherited values also when editing an existing entity, when the user changes the reference/parent

I think the last point is a rather common use case: A user creates an incomplete record, then goes back to edit and select a "parent" reference. This case should trigger the same logic editing an existing record as for editing a new record.
(This is a technical challenge of course, because it's difficult to detect if a field has been manually set/cleared before the last save, as we can't rely on form field dirty state ...)

Apart from this (important) edge case, seems to work great now! :+1:

**Review by sleidig — APPROVED** [review]

File: src/app/core/entity/schema/entity-schema-field.ts [hunk 4]
```diff
@@ -56,13 +57,33 @@ export interface EntitySchemaField {
   /**
    * Whether the field should be initialized with a default value if undefined
    * Default values are applied to form fields before they are displayed to users
+   *
+   * @deprecated replaced by defaultFieldValue
    */
   defaultValue?: PLACEHOLDERS | any;
```
[the remark this answers is not in the document]
Comment (tomwwinter) [reply src/app/core/entity/schema/entity-schema-field.ts]: I added the migrations and removed the deprecated code. But the latest changes may conflict with the changes from the array-refactoring. Maybe just merge this first: https://github.com/Aam-Digital/ndb-core/pull/2369

File: src/app/core/config/config.service.ts [hunk 5]
```diff
@@ -211,3 +216,32 @@ const migrateEntityDetailsInputEntityType: ConfigMigration = (
...
+  if (key !== "defaultValue") {
+    return configPart;
+  }
+
+  if (typeof configPart == "object") {
+    return configPart;
+  }
+
+  let placeholderValue: string | undefined = Object.values(PLACEHOLDERS).find(
+    (value) => value == configPart,
```
Comment (sleidig) [inline src/app/core/config/config.service.ts:233]: do we stick to JavaScript conventions of always using `===` comparison?
